File: stock-agent/apps/web/src/lib/schemas.ts
import { z } from "zod";

export const EvidenceItemSchema = z.object({
  type: z.enum(["news", "numeric", "note"]),
  title: z.string(),
  source: z.string(),
  published_at: z.string().nullable().optional(),
  snippet: z.string().nullable().optional(),
  doc_id: z.string().nullable().optional(),
});

export const AnalyzeResponseSchema = z.object({
  ticker: z.string(),
  signal: z.enum(["BUY", "SELL", "HOLD"]),
  confidence: z.number().min(0).max(1),
  horizon_days: z.number().int(),
  thesis_bullets: z.array(z.string()),
  numeric_features: z.record(z.number()),
  evidence: z.array(EvidenceItemSchema),
  risk: z.object({
    stop_loss_pct: z.number(),
    take_profit_pct: z.number(),
    invalidators: z.array(z.string()),
  }),
  provenance: z.object({
    tool_calls: z.array(z.record(z.any())),
    retrieved_doc_ids: z.array(z.string()),
    data_windows: z.record(z.any()),
  }),
});

export type AnalyzeResponse = z.infer<typeof AnalyzeResponseSchema>;

File: stock-agent/apps/api/app/agent/schemas.py
from __future__ import annotations
from typing import Literal, Any
from pydantic import BaseModel, Field

Signal = Literal["BUY", "SELL", "HOLD"]

class Constraints(BaseModel):
    max_position_pct: float = Field(default=0.10, ge=0.0, le=1.0)
    max_daily_loss_pct: float = Field(default=0.02, ge=0.0, le=1.0)

class AnalyzeRequest(BaseModel):
    ticker: str = Field(..., min_length=1, max_length=12)
    horizon_days: int = Field(default=20, ge=1, le=365)
    as_of: str | None = None  # ISO date string "YYYY-MM-DD" (optional for now)
    constraints: Constraints = Field(default_factory=Constraints)

class EvidenceItem(BaseModel):
    type: Literal["news", "numeric", "note"]
    title: str
    source: str
    published_at: str | None = None
    snippet: str | None = None
    doc_id: str | None = None

class RiskBlock(BaseModel):
    stop_loss_pct: float = Field(..., ge=0.0, le=1.0)
    take_profit_pct: float = Field(..., ge=0.0, le=3.0)
    invalidators: list[str] = Field(default_factory=list)

class Provenance(BaseModel):
    tool_calls: list[dict[str, Any]] = Field(default_factory=list)
    retrieved_doc_ids: list[str] = Field(default_factory=list)
    data_windows: dict[str, Any] = Field(default_factory=dict)

class AnalyzeResponse(BaseModel):
    ticker: str
    signal: Signal
    confidence: float = Field(..., ge=0.0, le=1.0)
    horizon_days: int
    thesis_bullets: list[str]
    numeric_features: dict[str, float] = Field(default_factory=dict)
    evidence: list[EvidenceItem] = Field(default_factory=list)
    risk: RiskBlock
    provenance: Provenance
